identity-iota/src/credential/credential_validator.rs

```diff
@@ -0,0 +1,645 @@
+    let is_ok = if let Some(expiration_date) = credential.expiration_date {
+      expiration_date >= timestamp
+    } else {
+      true
+    };
+    is_ok.then(|| ()).ok_or(ValidationError::ExpirationDate)
+  }
+
+  /// Validate that the [Credential] is issued no later than the specified [Timestamp].
+  ///
+  /// # Terminology
+  /// This is a *validation unit*
+  pub fn was_not_issued_after<T>(credential: &Credential<T>, timestamp: Timestamp) -> Result<()> {
+    Self::was_not_issued_after_local_error(credential, timestamp).map_err(Error::UnsuccessfulValidationUnit)
+  }
+
+  fn was_not_issued_after_local_error<T>(credential: &Credential<T>, timestamp: Timestamp) -> ValidationUnitResult {
+    (credential.issuance_date <= timestamp)
+      .then(|| ())
+      .ok_or(ValidationError::IssuanceDate)
+  }
+
+  // helper method for verifying the signature of a trusted issuer
+  fn extract_corresponding_issuer_helper<'a, T, I: Iterator<Item = &'a ResolvedIotaDocument>>(
+    credential: &Credential<T>,
+    mut issuer_docs: I,
+  ) -> std::result::Result<&'a ResolvedIotaDocument, ValidationError> {
+    let issuer_did: Result<IotaDID> = credential.issuer.url().as_str().parse();
+    match issuer_did {
+      Ok(did) => {
+        // if the issuer_did corresponds to one of the trusted issuers we use the corresponding DID Document to verify
+        // the signature
+        issuer_docs
+          .find(|issuer_doc| issuer_doc.document.id() == &did)
+          .ok_or(ValidationError::IncompatibleIssuerDocuments)
+      }
+      Err(error) => {
+        // the issuer's url could not be parsed to a valid IotaDID
+        Err(ValidationError::IssuerUrl { source: error.into() })
+      }
+    }
+  }
+  /// Verify the signature using the DID Document of a trusted issuer.
+  ///
+  /// This method will only attempt to verify the credential's signature if
+  /// the credential issuer url can be parsed to a DID belonging to one of the trusted issuers.
+  ///
+  /// # Terminology
+  /// This method is a *validation unit*
+  pub fn verify_signature<T: Serialize>(
+    credential: &Credential<T>,
+    trusted_issuers: &[ResolvedIotaDocument],
+    options: &VerifierOptions,
+  ) -> Result<()> {
+    Self::verify_signature_local_error(credential, trusted_issuers, options).map_err(Error::UnsuccessfulValidationUnit)
+  }
+
+  fn verify_signature_local_error<T: Serialize>(
+    credential: &Credential<T>,
+    trusted_issuers: &[ResolvedIotaDocument],
+    options: &VerifierOptions,
+  ) -> ValidationUnitResult {
+    Self::extract_corresponding_issuer_helper(credential, trusted_issuers.iter()).and_then(|issuer| {
+      issuer
+        .document
+        .verify_data(credential, options)
+        .map_err(|error| ValidationError::IssuerProof { source: error.into() })
+    })
+  }
+
+  /// Validates a [Credential].
+  ///
+  /// Common concerns are checked such as the credential's signature, expiration date, issuance date and semantic
+  /// structure.
+  ///
```

Comment:
It might be worth also documenting what is _not_ validated by this, such as the schema and holder.

Reply:
I will try to reformulate this, but I don't think it is realistic to mention everything that is not validated as that list is very long. There are so many possible subject to holder relationships an application may want to validate, then there is also fitness for purpose, credentialStatus etc.

Reply:
It doesn't have to be exhaustive, just enough to suggest that there are many things this validation doesn't cover that developers need to check themselves.